File: Creational/Factory/Factory_C++/Factory2.cpp
// Factory Design Pattern - Creational Category

// Factory Definition
// A utility class that creates an instance of a class from a family of derived classes.

// Abstract Factory Definition
// A utility class that creates an instance of several families of classes. 
// It can also return a factory for a certain group.

// The Factory Design Pattern is useful in a situation that requires the creation 
// of many different types of objects, all derived from a common base type. 
// The Factory Method defines a method for creating the objects, which subclasses 
// can then override to specify the derived type that will be created. 
// Thus, at run time, the Factory Method can be passed a description of a desired 
// object (e.g., a string read from user input) and return a base class pointer 
// to a new instance of that object. 
// The pattern works best when a well-designed interface is used for the base class, 
// so there is no need to cast the returned object.

// Problem 
// We want to decide at run time what object is to be created based on some 
// configuration or application parameter. When we write the code, we do not know 
// what class should be instantiated.

// Solution 
// Define an interface for creating an object, but let subclasses decide which 
// class to instantiate. Factory Method lets a class defer instantiation to subclasses.

// In the following example, a factory method is used to create laptop or desktop computer objects at run time.

// http://en.wikibooks.org/wiki/C%2B%2B_Programming/Code/Design_Patterns#Factory

#include <map>
#include <string>
#include <iostream>

using namespace std;

class Computer
{
	public:

		// Without this destructor, you do not call Laptop or Desktop destructor in this example!
		virtual ~Computer() {};

		virtual void Run() = 0;
		virtual void Stop() = 0;
		virtual void Name() = 0;
	};

	class Laptop : public Computer
	{
		public:

			// Because we have virtual functions, we need virtual destructor
			virtual ~Laptop() {};

			virtual void Run() { mHibernating = false; }; 
			virtual void Stop() { mHibernating = true; };
			virtual void Name() { cout << "I am a Laptop\n"; };

		private:

			// Whether or not the machine is hibernating
			bool mHibernating; 
	};

	class Desktop : public Computer
	{
		public:
		
			virtual ~Desktop() { };

			virtual void Run() {mOn = true;}; 
			virtual void Stop() {mOn = false;};
			virtual void Name() { cout << "I am a Desktop\n"; };
			

		private:

			bool mOn; // Whether or not the machine has been turned on
 };

// The actual ComputerFactory class returns a Computer, given a real world description of the object.
class ComputerFactory
{
	public:
	
		static Computer * NewComputer(const string & description)
		{
			if(description == "laptop")
				return new Laptop;

			if(description == "desktop")
				return new Desktop;

			return NULL;
		}
};

int main( int argc, char **argv )
{
	Computer * computer = NULL;

	computer = ComputerFactory::NewComputer("laptop");
	computer->Name();

	computer = ComputerFactory::NewComputer("desktop");
	computer->Name();

	system("pause");

	return 0;
}

File: Creational/Factory/Factory_C#/Factory2.cs
// Factory Design Pattern - Creational Category

// Definition
// Define an interface for creating an object, but let subclasses decide which class
// to instantiate. Factory Method lets a class defer instantiation to subclasses.

// Participants
// The classes and objects participating in this pattern are:
//
// Product (Page)
// - Defines the interface of objects the factory method creates
//
// ConcreteProduct (SkillsPage, EducationPage, ExperiencePage)
// - Implements the Product interface
//
// Creator (Document)
// - Declares the factory method, which returns an object of type Product. 
// - Creator may also define a default implementation of the factory method 
//   that returns a default ConcreteProduct object. 
// - May call the factory method to create a Product object.
//
// ConcreteCreator  (Report, Resume)
// - Overrides the factory method to return an instance of a ConcreteProduct.

// The following real-world code demonstrates the Factory method offering 
// flexibility in creating different documents. The derived Document classes 
// Report and Resume instantiate extended versions of the Document class. 
// Here, the Factory Method is called in the constructor of the Document base class.

// http://www.dofactory.com/net/factory-method-design-pattern

using System;
using System.Collections.Generic;

// MainApp startup class for Real-World Factory Method Design Pattern.
class MainApp
{
    // Entry point into console application.
    static void Main()
    {
        // Note: constructors call Factory Method
        Document[] documents = new Document[2];

        documents[0] = new Resume();
        documents[1] = new Report();

        // Display document pages
        foreach (Document document in documents)
        {
            Console.WriteLine("\n" + document.GetType().Name + "--");
            foreach (Page page in document.Pages)
            {
                Console.WriteLine(" " + page.GetType().Name);
            }
        }

        // Wait for user
        Console.ReadKey();
    }
}

// The 'Product' abstract class
abstract class Page { }

// A 'ConcreteProduct' class
class SkillsPage : Page { }

// A 'ConcreteProduct' class
class EducationPage : Page { }

// A 'ConcreteProduct' class
class ExperiencePage : Page { }

// A 'ConcreteProduct' class
class IntroductionPage : Page { }

// A 'ConcreteProduct' class
class ResultsPage : Page { }

// A 'ConcreteProduct' class
class ConclusionPage : Page { }

// A 'ConcreteProduct' class
class SummaryPage : Page { }

// A 'ConcreteProduct' class
class BibliographyPage : Page { }

// The 'Creator' abstract class
abstract class Document
{
    private List<Page> _pages = new List<Page>();

    // Constructor calls abstract Factory method
    public Document()
    {
        this.CreatePages();
    }

    public List<Page> Pages
    {
        get { return _pages; }
    }

    // Factory Method
    public abstract void CreatePages();
}

// A 'ConcreteCreator' class
class Resume : Document
{
    // Factory Method implementation
    public override void CreatePages()
    {
        Pages.Add(new SkillsPage());
        Pages.Add(new EducationPage());
        Pages.Add(new ExperiencePage());
    }
}

// A 'ConcreteCreator' class
class Report : Document
{
    // Factory Method implementation
    public override void CreatePages()
    {
        Pages.Add(new IntroductionPage());
        Pages.Add(new ResultsPage());
        Pages.Add(new ConclusionPage());
        Pages.Add(new SummaryPage());
        Pages.Add(new BibliographyPage());
    }
}